internal/policy/policy.go

```diff
@@ -359,6 +359,111 @@ type Wrapper struct {
+	var actions []string
+	if p == nil {
+		return actions
+	}
+
+	lut := make(map[string]struct{})
+	switch p := p.PolicyType.(type) {
+	case *policyv1.Policy_ResourcePolicy:
+		for _, r := range p.ResourcePolicy.Rules {
+			for _, a := range r.Actions {
+				if _, ok := lut[a]; !ok {
+					lut[a] = struct{}{}
+					actions = append(actions, a)
+				}
+			}
+		}
+	case *policyv1.Policy_PrincipalPolicy:
+		for _, r := range p.PrincipalPolicy.Rules {
+			for _, a := range r.Actions {
+				if _, ok := lut[a.Action]; !ok {
+					lut[a.Action] = struct{}{}
+					actions = append(actions, a.Action)
+				}
+			}
+		}
+	}
+
+	sort.Strings(actions)
+	return actions
+}
+
+// ListVariables returns local and exported variables (not imported ones) defined in a policy.
+func ListVariables(p *policyv1.Policy) []*responsev1.InspectPoliciesResponse_Variable {
+	var variables []*responsev1.InspectPoliciesResponse_Variable
+	if p == nil {
+		return variables
+	}
+
+	switch pt := p.PolicyType.(type) {
+	case *policyv1.Policy_DerivedRoles:
+		if pt.DerivedRoles.Variables == nil {
+			return variables
+		}
+
+		for name, value := range pt.DerivedRoles.Variables.Local {
+			variables = append(variables, &responsev1.InspectPoliciesResponse_Variable{
+				Name:  name,
+				Value: value,
+				Source: &responsev1.InspectPoliciesResponse_Variable_Source{
+					Type: responsev1.InspectPoliciesResponse_Variable_Source_TYPE_LOCAL,
+					Id:   namer.PolicyKey(p),
+				},
+			})
+		}
+	case *policyv1.Policy_ExportVariables:
+		for name, value := range pt.ExportVariables.Definitions {
+			variables = append(variables, &responsev1.InspectPoliciesResponse_Variable{
+				Name:  name,
+				Value: value,
+				Source: &responsev1.InspectPoliciesResponse_Variable_Source{
+					Type: responsev1.InspectPoliciesResponse_Variable_Source_TYPE_EXPORTED,
+					Id:   namer.PolicyKey(p),
+				},
+			})
+		}
+	case *policyv1.Policy_PrincipalPolicy:
+		if pt.PrincipalPolicy.Variables == nil {
+			return variables
+		}
+
+		for name, value := range pt.PrincipalPolicy.Variables.Local {
+			variables = append(variables, &responsev1.InspectPoliciesResponse_Variable{
+				Name:  name,
+				Value: value,
+				Source: &responsev1.InspectPoliciesResponse_Variable_Source{
+					Type: responsev1.InspectPoliciesResponse_Variable_Source_TYPE_LOCAL,
+					Id:   namer.PolicyKey(p),
+				},
+			})
+		}
+	case *policyv1.Policy_ResourcePolicy:
+		if pt.ResourcePolicy.Variables == nil {
+			return variables
+		}
+
+		for name, value := range pt.ResourcePolicy.Variables.Local {
+			variables = append(variables, &responsev1.InspectPoliciesResponse_Variable{
+				Name:  name,
+				Value: value,
+				Source: &responsev1.InspectPoliciesResponse_Variable_Source{
+					Type: responsev1.InspectPoliciesResponse_Variable_Source_TYPE_LOCAL,
+					Id:   namer.PolicyKey(p),
+				},
+			})
+		}
+	}
+
+	sort.Slice(variables, func(i, j int) bool {
+		return sort.StringsAreSorted([]string{variables[i].Name, variables[j].Name})
```

Comment:
Isn't it more efficient to do `variables[i].Name < variables[j].Name`? Otherwise this requires allocating an array and calling a function, which may or may not get inlined.